# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 0304 Гурьянов Савелий, вариант 31.

## Цель работы

Сформировать представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

Аддитивной цепочкой для $ n \in ℕ $ называется последовательность натуральных чисел
$$ 1 = a_0, a_1, ..., a_r = n, $$
где каждый элемент последователньости равен сумме каких-то двух предыдущих:
$$ a_i = a_j + a_k, \quad k \le j \le i, \quad i = 1, 2, ..., r $$

$l(n) = r$ - наименьшая длина аддитивной цепочки для $ n \in ℕ $.

Для метода наименьших множителей: $ \quad l(mn) \le l(m) + l(n) $

Для m-арного метода: $ \quad l(n) \le m - 2 + (k + 1)t \quad [m = 2^k, n = \sum_{j = 0}^t d_j m^{t-j}] $

Для SX-метода: $ \quad l(n) \le \lambda (n) + \nu (n) - 1 $


### Свойства аддитивных цепочек:
* Полагается строгое возрастание элементов цепочки:
$ 1 = a_0 < a_1 < a_2 < ... < a_n = n $
* Одинаковые числа в цепочке можно опустить
* Пара $ (j, k), 0 \le k \le j < i $ называется шагом $i$
* Если $\exists$ более чем 1 пара $ (j, k) $, полагаем $ max \hspace{0.2cm} j $

### Виды шагов:
* удвоение: $ j = k = i - 1 $
* звёздный шаг: $ j = i - 1 $ (линейный шаг)
* малый шаг: $ \lambda (a_i) = \lambda (a_i - 1) $, где $ \lambda (n) = \lfloor lb(n) \rfloor$

### Свойства видов шагов:
* шаг 1 - всегда удвоение
* удвоение - звёздный шаг, но никогда не малый
* если $i$-ый шаг не малый, то $(i+1)$-ый шаг либо малый, либо звёздный, либо оба
* за удвоением всегда следует звёздный шаг
* если $(i+1)$-ый шаг не звёздный и не малый, то $i$-ый шаг должен быть малым

### Теорема:

Если аддитивная цепочка содержит $d$ и $f = r - d$ неудвоений, то $n \le 2^{d-1} F_{f+3}$, где $F_j$ - число Фиббоначи

### Следствие:

Если аддитивная цепочка содержит $f$ удвоений и $S$ малых шагов, то

${S \le f \le} {S \over {1 - lb(\varphi)}}$, где $\varphi = {{\sqrt{5} + 1} \over 2} $ - золотое сечение

### Алгоритм Брауэра:

Для $n \in ℕ$ при заданном $k \in ℕ$ можно построить цепочку Брауэра с помощью рекуррентной формулы:

$$ B_k (n) =
\begin{cases}
1, 2, 3, ..., 2^k - 1, \quad n < 2^k \\
B_k (q), 2q, 4q, ..., 2^k q, n, \quad n \ge 2^k
\end{cases} \\
q = \lfloor {n \over 2^k} \rfloor
$$

Данная цепочка имеет длину
$$l_B (n) = j(k + 1) + 2^k - 2,$$
при условии что $jk \le \lambda (n) \le (j+1)k$

Длина будет минимализирована для больших $n$, если положить $k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n)$

**Ход алгоритма:**

* Задаётся некий параметр $k$ для $n$.
Выполняется вычисление вспомогательных чисел:
$$
d = 2^k, \hspace{0.2cm} q_1 = [ {n \over d} ], \hspace{0.2cm} r_1 = n \hspace{0.2cm} mod \hspace{0.2cm} d => n = q_1 d + r_1 \quad (0 \le r_1 < d) \\
q_2 = [ {q_1 \over d} ], \hspace{0.2cm} r_2 = q_1 \hspace{0.2cm} mod \hspace{0.2cm} d => q_1 = q_2 d + r_2 \quad (0 \le r_2 < d)
$$
* Данная процедура продолжается, пока не появится $q_s < d,$ следовательно $q_{s-1} = q_s d + r_s$
* Таким образом, n имеет вид
$$ n = 2^k q_1 + r_1 = 2^k (2^k q_2 + r_2) + r_1 = ...\\
... = 2^k (2^k (... (2^k q_s + r_s ) ... ) + r_2 ) + r_1 . $$

$$B_n (n): 1, 2, 3, ..., 2^k - 1, \\
2q_s, 4q_s, 8q_s, ..., 2^k q_s, 2^k q_s + r_s, \\
2q_{s-1}, 4q_{s-1}, 8q_{s-1}, ..., 2^k q_{s-1}, 2^k q_{s-1} + r_{s-1}, \\
..., \\
..., 2^k q_1, 2^k q_1 + r_1 = n.$$

### Алгоритм Яо:

* Обладает такой же вычислительной мощностью, что и алгоритм Брауэра
* Начало похоже: $k \ge 2$, $n$ раскладывается в $2^k$-ой системе счисления:
$$ n = \sum_{i = 0}^j a_i 2^{ik} , a_j \ne 0 $$
* Введём функцию d:
$$ d(z) = \sum_{i: a_i = z} 2^{ik} $$

**Ход алгоритма:**

* Базовая последовательность: $1, 2, 4, ..., 2^{\lambda(n)}$
* Вычисление значений $d(z)$ для всех $z \in \{ 1, 2, ..., 2^k - 1\}, \quad d(z) \ne 0$
* Вычисление $zd(z)$ для всех $z$
* n раскладывается в виде
$$ n = \sum_{z = 1}^{2^k - 1} zd(z) $$

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы

1. Вручную (т.е. не реализовывая алгоритм на Sage) построить последовательность вычислений бинарным методом и методом множителей для $x^n$ для 2-3 значений $n$ ($n \ge 30$). Сравнить количество операций для каждого метода, сделать выводы.

**n = 90:**<br>
Двоичный метод:<br>

N | Y | Z
--- | --- | ---
90 | 1 | x
45 | 1 | x^2
22 | x^2 | x^4
11 | x^2 | x^8
5 | x^10 | x^16
2 | x^26 | x^32
1 | x^26 | x^64
0 | x^90 | x^128

Метод множителей:<br>

$$ 90 = pq = 2 \cdot 45 $$
$$ y = x^2\\
   y^{45} = (y^3)^{15} = ((y^3)^3)^5 = ((y * y^2)^3)^5
$$
   
**n = 99:**<br>
Двоичный метод:<br>

N | Y | Z
--- | --- | ---
99 | 1 | x
49 | x | x^2
24 | x^3 | x^4
12 | x^3 | x^8
6 | x^3 | x^16
3 | x^3 | x^32
1 | x^35 | x^64
0 | x^99 | x^128

Метод множителей:<br>

$$ 99 = pq = 3 \cdot 33 $$
$$ y = x^3\\
   y^{33} = (y^{11})^3 = (y^{10} \cdot y )^3 = ((y^5)^2 \cdot y)^3 = ((y^4 \cdot y)^2 \cdot y)^3\\
   y^{33} = ((((y^2)^2) \cdot y)^2 \cdot y)^3
$$
Количество операций:<br>

n | двоичный метод | метод множителей
--- | --- | ---
90 | 10 | 5
99 | 10 | 7


Чем больше единиц в двоичной записи числа, тем медленнее работает двоичный метод. Поэтому в числах с большим количеством единиц в двочиной записи данный метод работает достаточно медленно. Таким образом, к числам с большим количеством единиц в двоичной записи лучше применять метод

2. Реализовать алгоритм Брауэра для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра k, сопоставить длины полученных аддитивных цепочек с минимальной аддитивной цепочкой для заданного числа. Сделать выводы.

In [9]:
def b_len(x): #Возвращает "количество цифр в двоичной записи - 1"
	return Integer(x).nbits() - 1

def minimum_length(n): #Минимальная длина цепочки
	return int(b_len(n) + b_len(n) / b_len(b_len(n)) + (b_len(n) * b_len(b_len(b_len(n)))) / (b_len(b_len(n)) ** 2))


def delete_dublicates(old_chain): #Удаляет дубликаты из цепи, преобразуя список в словарь и обратно
	new_chain = list(dict.fromkeys(old_chain))
	if new_chain[-1] == 0:
		del new_chain[-1]
	return new_chain

def make_str(chain): #Создаёт строковое представление цепи
	return " ".join(map(str, chain[:len(chain)]))


def algorithm(n, k):
	n = Integer(n)

	d = 2 ** k

	remainders = [] # массив остатков
	q = n.quo_rem(d)[0]
	remainders.append(n.quo_rem(d)[1])
	sqmax = remainders[0]

	i = 0
	while q >= d:
		remainders.append(q.quo_rem(d)[1])
		q = q.quo_rem(d)[0]
		if remainders[i] > sqmax:
			sqmax = remainders[i]
		i += 1
	remainders.append(q)
	sqmax = max(q, sqmax)

	chain = [0] * (sqmax + (len(remainders) - 1) * (k + 1))
	for i in range(0, sqmax):
		chain[i] = i + 1

	chain[sqmax - 1] = q
	for i in range(0, len(remainders) - 1):
		for j in range(0, k):
			chain[sqmax + i * (k + 1) + j] = chain[sqmax + i * (k + 1) + j - 1] * 2
		chain[sqmax + i * (k + 1) + k] = chain[sqmax + i * (k + 1) + k - 1] + remainders[len(remainders) - 2 - i]
	chain[sqmax - 1] = sqmax

	return delete_dublicates(chain)


set_random_seed(15623)
n_set = []
for i in range(0, 3):
	n_set.append(randint(50, 100))
for i in range(0, 3):
	n_set.append(randint(200, 50000))

for k in [2, 4, 6, 8]:
	print("k = ", k, ":", sep = "")
	for n in n_set:
		chain = algorithm(n, k)
		print("n = {}; l(n) = {}; оптимальная l(n) = {}; цепочка: {}".format(n, len(chain), minimum_length(n), make_str(chain)))


k = 2:
n = 80; l(n) = 8; оптимальная l(n) = 10; цепочка: 1 2 4 5 10 20 40 80
n = 89; l(n) = 10; оптимальная l(n) = 10; цепочка: 1 2 4 5 10 20 22 44 88 89
n = 75; l(n) = 10; оптимальная l(n) = 10; цепочка: 1 2 3 4 8 16 18 36 72 75
n = 25995; l(n) = 22; оптимальная l(n) = 20; цепочка: 1 2 3 4 6 12 24 25 50 100 101 202 404 406 812 1624 3248 6496 6498 12996 25992 25995
n = 46217; l(n) = 21; оптимальная l(n) = 21; цепочка: 1 2 4 8 11 22 44 45 90 180 360 720 722 1444 2888 5776 11552 11554 23108 46216 46217
n = 28066; l(n) = 22; оптимальная l(n) = 20; цепочка: 1 2 3 4 6 12 24 27 54 108 109 218 436 438 876 1752 1754 3508 7016 14032 28064 28066
k = 4:
n = 80; l(n) = 9; оптимальная l(n) = 10; цепочка: 1 2 3 4 5 10 20 40 80
n = 89; l(n) = 14; оптимальная l(n) = 10; цепочка: 1 2 3 4 5 6 7 8 9 10 20 40 80 89
n = 75; l(n) = 15; оптимальная l(n) = 10; цепочка: 1 2 3 4 5 6 7 8 9 10 11 16 32 64 75
n = 25995; l(n) = 26; оптимальная l(n) = 20; цепочка: 1 2 3 4 5 6 7 8 9 10 11 12 24 48 96 101 202 404 808 

Видно, что по мере увеличения k цепочка теряет оптимальность. Особенно хорошо это заметно для больших n. К примеру, для n = 28066. Поэтому для наибольшей оптимальности нужно, чтобы k было гораздо меньше n. Точная формула, по которой это определяется: $ k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n) $.

3. Реализовать алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестировать алгоритм минимум для 5 значений n > 1000. Указать, сколько времени потребовалось на поиск цепочки и какая цепочка получилась. Сравнить с предыдущими методами, сделать выводы.

In [4]:
import math
def b_len(x):
	return Integer(x).nbits() - 1

def b_count(x):
	return Integer(x).popcount()


def get_min_chain_length(x):
	return b_len(x) + 1

def make_chain(vector): #создать цепочку из вектора
	numbers = [0] * (len(vector) + 1)
	numbers[0] = 1

	for i in range(0, len(vector)):
		numbers[i + 1] = numbers[i] + numbers[vector[i] - 1]
	return numbers

def make_str(chain):
	return " ".join(map(str, chain))


def create_vector(n, s = 1):
	vector = [0] * n
	for i in range(0, n):
		vector[i] = i + s
	return vector

def check_minimum(vector):
	for i in vector:
		if i > 1:
			return False
	return True

def dec_vector(vector):

	for i in range(0, len(vector)):
		if vector[len(vector) - 1 - i] > 1:
			vector[len(vector) - 1 - i] -= 1
			for j in range(len(vector) - i, len(vector)):
				vector[j] = j + 1
			return

def get_last_number(vector):
	numbers = [0] * (len(vector) + 1)
	numbers[0] = 1

	for i in range(0, len(vector)):
		numbers[i + 1] = numbers[i] + numbers[vector[i] - 1]
	return numbers[len(vector)]


def algorithm_crush_vector(n):
	tries = [1, 1]

	lmin = math.ceil(math.log(n, 2))
	lmax = b_len(n) + b_count(n) - 1

	for m in range(lmin, lmax + 1):
		q = int(m / 2)
		iv_fix = create_vector(q)

		while True:
			iv_mut = create_vector(m - q, q + 1)

			iv_fix.append(iv_mut[0])
			aq = get_last_number(iv_fix)
			del iv_fix[-1]
			amin = 1 + m - q
			amax = aq * 2 ** (m - q)

			if n > amax or n < amin:
				if check_minimum(iv_fix):
					break
				dec_vector(iv_fix)
				continue

			while True:
				if get_last_number(iv_fix + iv_mut) == n:
					return (iv_fix + iv_mut, tries)

				if check_minimum(iv_mut):
					break
				dec_vector(iv_mut)
				tries[1] += 1

			if check_minimum(iv_fix):
				break
			dec_vector(iv_fix)
			tries[0] += 1


set_random_seed(15623)
n_set = []
for i in range(0, 5):
	n_set.append(randint(1001, 1250))
print("{}\n".format(n_set))
for n in n_set:
	res = algorithm_crush_vector(n)
	print("n = {}  len = {}; min len = {}; перебрано фиксированных частей: {:6}   перебранно изменяемых частей: {:6} \n{}\n".format(n, len(res[0]), get_min_chain_length(n), res[1][0], res[1][1], make_str(make_chain(res[0]))))


[1123, 1157, 1104, 1101, 1180]

n = 1123  len = 13; min len = 11; перебрано фиксированных частей:    825   перебранно изменяемых частей: 6772542 
1 2 4 8 16 32 34 68 136 272 544 1088 1122 1123

n = 1157  len = 12; min len = 11; перебрано фиксированных частей:    364   перебранно изменяемых частей: 2838914 
1 2 4 5 9 18 36 72 144 288 576 1152 1157

n = 1104  len = 12; min len = 11; перебрано фиксированных частей:    107   перебранно изменяемых частей: 731856 
1 2 4 8 16 32 64 128 256 512 1024 1088 1104

n = 1101  len = 13; min len = 11; перебрано фиксированных частей:    878   перебранно изменяемых частей: 9793449 
1 2 4 8 12 13 17 34 68 136 272 544 1088 1101

n = 1180  len = 13; min len = 11; перебрано фиксированных частей:    801   перебранно изменяемых частей: 6689765 
1 2 4 8 16 24 48 96 192 384 768 1152 1176 1180



Чем ближе число к степени двойки, тем меньше необходимо перебрать векторов индексов. При этом длина полученной цепочки уже теряет зависимость от n.

Данный алгоритм имеет среднюю сложность $O(\sum(m-1)!)$, где $m$ - длина минимальной звёздной цепочки.

In [17]:
import math
import random

def b_len(x):
	return Integer(x).nbits() - 1

def b_count(x):
	return Integer(x).popcount()


def get_min_chain_length(x):
	return b_len(x) + 1

def make_chain(vector): #создать цепочку из вектора
	numbers = [0] * (len(vector) + 1)
	numbers[0] = 1

	for i in range(0, len(vector)):
		numbers[i + 1] = numbers[i] + numbers[vector[i] - 1]
	return numbers

def make_str(chain):
	return " ".join(map(str, chain))


def create_vector(n, s = 1):
	vector = [0] * n
	for i in range(0, n):
		vector[i] = i + s
	return vector

def check_minimum(vector):
	for i in vector:
		if i > 1:
			return False
	return True

def dec_vector(indvec):
	indvec_len = len(indvec)

	for i in range(0, indvec_len):
		if indvec[indvec_len - 1 - i] > 1:
			indvec[len(indvec) - 1 - i] -= 1
			for j in range(indvec_len - i, indvec_len):
				indvec[j] = j + 1
			return

def get_last_number(vector):
	numbers = [0] * (len(vector) + 1)
	numbers[0] = 1

	for i in range(0, len(vector)):
		numbers[i + 1] = numbers[i] + numbers[vector[i] - 1]
	return numbers[len(vector)]


def algorithm_crush_vector(n):
	tries = [1, 1]

	lmin = math.ceil(math.log(n, 2))
	lmax = b_len(n) + b_count(n) - 1

	for m in range(lmin, lmax + 1):
		q = int(m / 2)
		iv_fix = create_vector(q)

		while True:
			iv_mut = create_vector(m - q, q + 1)

			iv_fix.append(iv_mut[0])
			aq = indvec_get_last_num(iv_fix)
			del iv_fix[-1]
			amin = 1 + m - q
			amax = aq * 2 ** (m - q)

			if n > amax or n < amin:
				if check_minimum(iv_fix):
					break
				dec_vector(iv_fix)
				continue

			while True:
				if indvec_get_last_num(iv_fix + iv_mut) == n:
					return (iv_fix + iv_mut, tries)

				if check_minimum(iv_mut):
					break
				dec_vector(iv_mut)
				tries[1] += 1

			if check_minimum(iv_fix):
				break
			dec_vector(iv_fix)
			tries[0] += 1


print("l*(2^n-1) <= l*(n) +  n - 1")
for n in range(2, 13):
	res = algorithm_crush_vector(n)
	resb = algorithm_crush_vector(2 ** n - 1)
	print("{:9} <= {:5} + {:2} - 1\n{}".format(len(resb[0]), len(res[0]), n, len(resb[0]) <= len(res[0]) + n - 1))


l*(2^n-1) <= l*(n) +  n - 1
        2 <=     1 +  2 - 1
True
        4 <=     2 +  3 - 1
True
        5 <=     2 +  4 - 1
True
        7 <=     3 +  5 - 1
True
        8 <=     3 +  6 - 1
True
       10 <=     4 +  7 - 1
True
       10 <=     3 +  8 - 1
True
       12 <=     4 +  9 - 1
True
       13 <=     4 + 10 - 1
True


Результаты проверки гипотезы Шольца-Брауэра для звёздных цепочек предоставлены в следующей таблице:

$l^*(2^n - 1)$| $l^*(n)$ | $n$ | $l^*(2^n - 1) \le l^*(n) + n - 1$
--- | --- | --- | ---
3 | 2 | 2 | +
5 | 3 | 3 | +
6 | 3 | 4 | +
8 | 4 | 5 | +
9 | 4 | 6 | +
11 | 5 | 7 | +
11 | 4 | 8 | +
13 | 5 | 9 | +
14 | 5 | 10 | +
16 | 6 | 11 | +
16 | 5 | 12 | +

Таким образом, было подтверждено, что гипотеза Шольца-Брауэра для звёздных цепочек верна для $2 \le n \le 12$ (для $1$ очевидно $1 \le 1 + 1 - 1$)

### Выводы
При выполнении практической работы был изучен алгоритм Браэура, получены навыки работы с SageMath. На примере нескольких чисел построены аддитивные цепочки с помощью бинарного метода и метода множителей. На практике проверена гипотеза Шольца-Браэура.